.github/workflows/auto-merge.yml

```diff
@@ -0,0 +1,115 @@
+name: Automerge PRs
+
+on:
+  schedule:
+    - cron: "*/10 * * * *" # Run every 10 minutes
+
+jobs:
+  merge:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Automerge PRs with "automerge" label created within the last month and are mergeable
+        uses: actions/github-script@v6
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          retries: 3
+          script: |
+            const { repo: { owner, repo } } = context;
+
+            const MERGE_INTERVAL_MS = 5000;   // 5 seconds pause after a merge
+
+            async function sleep(ms) {
+              return new Promise(resolve => setTimeout(resolve, ms));
+            }
+
+            async function logRateLimit() {
+              const { data: rateLimit } = await github.rest.rateLimit.get();
+              console.log(`Rate limit remaining: ${rateLimit.resources.core.remaining}`);
+              console.log(`Rate limit resets at: ${new Date(rateLimit.resources.core.reset * 1000).toISOString()}`);
+            }
+
+            async function fetchPullRequestDetails(prNumber) {            
+              const pullRequest = await github.rest.pulls.get({
+                owner,
+                repo,
+                pull_number: prNumber
+              }).then(res => res.data);
+
+              if (pullRequest.mergeable !== null) {
+                return pullRequest;
+              }
+              return null;
+            }
+
+            async function isPRApproved(prNumber) {
+              const { data: reviews } = await github.rest.pulls.listReviews({
+                owner,
+                repo,
+                pull_number: prNumber
+              });
+              return reviews.some(review => review.state === 'APPROVED');
+            }
+
+            async function areAllChecksPassed(sha) {
+              const { data: checkRuns } = await github.rest.checks.listForRef({
+                owner,
+                repo,
+                ref: sha
+              });
+              return checkRuns.check_runs.every(run => run.conclusion === 'success');
+            }
+
+            // Get date from a month ago in ISO format
+            const oneMonthAgo = new Date();
+            oneMonthAgo.setMonth(oneMonthAgo.getMonth() - 1);
+            const sinceDate = oneMonthAgo.toISOString();
+
+            // List PRs with the "automerge" label created within the last month
+            const { data: issues } = await github.rest.issues.listForRepo({
+              owner,
+              repo,
+              labels: 'automerge',
+              since: sinceDate
+            });
+
+            // Filter for pull requests from the list of issues
+            const pullRequests = issues.filter(issue => issue.pull_request);
+
+            for (const pr of pullRequests) {
```

Comment:
Can we validate the PR is filed by a maintainer to prevent malicious code from being merged?

Reply:
That's a valid point. Can I confirm that you meant to limit this functionality only to the PRs of maintainers?

Reply:
Yes, that's what I meant. Another option is to remove the automerge label or check the last commit is made before an approval.

Reply:
I'm leaning towards `no commit is pushed after approval` because this also allows us to automerge PRs filed from external contributors.

Reply:
I agree to disallow the new commits after approval. Then we probably just need to add "Dismiss stale pull request approvals when new commits are pushed" to the branch protection rule.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule

Reply:
I think this would also block maintainers.

Reply:
I mean, can't we have the approval check only to no-maintainers and add the branch protection policy?

Reply:
I think we should always check the approval.

Reply:
Got it, so the expected behavior is like this?

- Maintainers
  - Need approval to merge the PR
  - Don't need to get approval again after adding new commits
- Non-maintainers
  - Need approval to merge the PR
  - Need to get approval again after adding new commits

If so, probably having another workflow that removes the auto-merge label when new commits are added would be the appropriate way since it can fulfill the requirement while keeping the behavior close to the Github merge-timer feature.

Reply:
Correct. +1 for removing the automerge label.

Reply:
I've added the new workflow. This workflow was verified in https://github.com/TomeHirata/mlflow_10069/pull/4